Tidy up documentation, add simple examples for illustration

CTrabant opened the pull request:
Clean up the documentation: fixing errors, clarifying the function of each method and return values consistently and adding simple examples.

CTrabant commented:
Addresses most of #70

manchicken commented on RabbitMQ.pm:
```diff
@@ -81,26 +83,30 @@ will exit with error code 141
 =head2 get_server_properties()
 
-Get a hashref of server properties (these may vary, you should C<Data::Dumper> to inspect). They will be provided by the RabbitMQ server to which you are connected.
+Get a reference to hash of server properties (these may vary, you should use C<Data::Dumper> to inspect). Properties will be provided for the RabbitMQ server to which you are connected.
```
Hashref is a specific technical term, I think we should keep that word in the docs since it adds clarity.

CTrabant replied:
I added (hashref) back to the docs.  I'm a little confused how hashref adds clarity over "reference to hash" so I split the difference left them both in.

manchicken replied:
Yeah, so, folks who program Perl know exactly what you mean by `hashref`, and they will also know what you mean by _reference to hash_. However, consider the following two scenarios:

A:

```perl
my $foo = { a => 1 };
bar( $foo );
```

and B:

```perl
my %foo = ( a => 1 );
bar( \%foo );
```

In _A_, you never see the hash. For me, that is a scalar, but it is also a hashref. In _B_, that's a reference to a hash because I have a hash and I'm passing a reference. Does that make sense?

Then there's the linguistics of it, if you can say the same thing two different ways, prefer the shorter way of saying it.

CTrabant replied:
I tried re-writing the reference-using sentences and just couldn't do it.  Even the perlref tutorial does not use `hashref` as a noun, only a variable name as an apparent contraction for "hash reference".

As for linguistics, the editor in me says the places in the doc would be (were) wrong if `hashref` were swapped out for what I wrote because they all refer to what the hash contains.  What I mean is "Get a hashref of server properties" is wrong because the hashref does not have server properties, the hash it refers to does.  Likewise, "This method returns a hashref containing the following information:" is wrong because the hashref does not contain the information, the hash it refers to does.  They are just references, they do not _have_ or _contain_ anything, they _are_ a pointer/link/whatever to something else.

Thanks for your examples, I get the subtle difference and I get that hashref is in common usage.  I just think that hashref is a colloquial contraction for "hash reference", which is just a more concise way to write "reference to hash" and that there is not really anything magical or more informative about it.

If this is a blocker for merging I'll change it but I'll rewrite the sentences so that hashrefs are only references without stuff in them.

manchicken commented on RabbitMQ.pm:
```diff
@@ -81,26 +83,30 @@ will exit with error code 141
 
 =head2 channel_open($channel)
 
+Open an AMQP channel to the RabbitMQ server.
```
This is more opening a channel on a connection than to the server.

manchicken commented on RabbitMQ.pm:
```diff
@@ -109,6 +115,8 @@ Returns the maximum allowed channel number.
 
 =head2 exchange_declare($channel, $exchange, $options, $arguments)
 
+Declare an AMQP exchange on the RabbitMQ server.
```
Maybe add "unless it already exists"? Maybe add stuff about how this goes boom if the name exists with different parameters?

manchicken commented on RabbitMQ.pm:
```diff
@@ -192,16 +216,10 @@ C<$arguments> is an optional hash which will be passed to the server
-
 =head2 queue_bind($channel, $queuename, $exchange, $routing_key, $arguments)
 
+Bind the specified queue to the specified exchange with a routing key.
```
More than just a routing key is supported. Maybe "given routing key and/or parameters"?

manchicken commented on RabbitMQ.pm:
```diff
@@ -231,10 +249,10 @@ C<$options> is an optional hash respecting the following keys:
-
 =head2 publish($channel, $routing_key, $body, $options, $props)
 
+Send a message to the RabbitMQ server.
```
Publish a message to a given exchange

manchicken commented on RabbitMQ.pm:
```diff
@@ -275,6 +293,12 @@ C<$props> is an optional hash (the AMQP 'props') respecting the following keys:
+Notify the RabbitMQ server that messages for the specified queue should
+be delivered on the specified channel.
+
+The consumer_tag is returned.  This command does B<not> return AMQP
```
`C<consumer_tag>`

manchicken commented on RabbitMQ.pm:
```diff
@@ -275,6 +293,12 @@ C<$props> is an optional hash (the AMQP 'props') respecting the following keys:
+be delivered on the specified channel.
+
+The consumer_tag is returned.  This command does B<not> return AMQP
+frames, for that the C<recv> method should be used.
```
`recv()` doesn't return frames, it returns complete messages.

CTrabant replied:
Fixed.  I also removed all mentions of frames, which do not seem important at this interface level.

manchicken commented on RabbitMQ.pm:
```diff
@@ -275,6 +293,12 @@ C<$props> is an optional hash (the AMQP 'props') respecting the following keys:
 =head2 consume($channel, $queuename, $options)
 
+Notify the RabbitMQ server that messages for the specified queue should
+be delivered on the specified channel.
```
This should probably just say that it puts a channel into consume mode. You don't actually tell the server anything other than that, really. You can enter consume mode and do nothing, and then die, and the server won't have done anything but ack your RPC calls.

manchicken commented on RabbitMQ.pm:
```diff
@@ -288,20 +312,11 @@ C<$options> is an optional hash respecting the following keys:
-wait for a message. If you do not provide a timeout (or set it to 0), then
-this call will block until it receives a message. If you set it to -1 it will
-return immediately (waiting 0 ms).
+Receive AMQP frames.
```
Messages, not frames.

manchicken commented on RabbitMQ.pm:
```diff
@@ -288,20 +312,11 @@ C<$options> is an optional hash respecting the following keys:
      }
 
-
-The consumer_tag is returned.  This command does B<not> return AMQP
```
This was inaccurate before.

manchicken commented on RabbitMQ.pm:
```diff
@@ -330,25 +345,32 @@ C<$props> is the hash sent by publish()  respecting the following keys:
 =head2 cancel($channel, $consumer_tag)
 
+This method returns true or false indicating whether we got the expected
+"cancel-ok" response from the server.
```
Maybe a line about what the method does?

CTrabant replied:
I have no idea what `cancel` does!  Suggestion?

manchicken replied:
Sure! The `cancel` call actually pulls a channel out of `consume` mode. More docs are available here: https://www.rabbitmq.com/consumer-cancel.html

CTrabant replied:
Thanks.  I've added a description for this method.

manchicken commented on RabbitMQ.pm:
```diff
@@ -429,24 +457,26 @@ C<$requeue> specifies if the message should be requeued.
-
 C<$channel> is a channel that has been opened with C<channel_open>.
 
-Commit a server-side (tx) transaction over $channel.
```
Removing the description?

manchicken replied:
Ack, ignore comment. Misread.

manchicken commented on RabbitMQ.pm:
```diff
@@ -479,6 +507,54 @@ the connection has been closed by the remote host, or otherwise invalidated,
 the socket will also be closed and should be re-opened before any additional
 calls are made.
 
+=head1 EXAMPLES
```
These are great!

manchicken commented on RabbitMQ.pm:
```diff
@@ -479,6 +507,54 @@ the connection has been closed by the remote host, or otherwise invalidated,
+    use Net::AMQP::RabbitMQ;
+    use Data::Dumper;
+
+    my $channel = 1;
```
Channel as variable, but you still use the constant `1` everywhere.

CTrabant replied:
Fixed, duh.  That's what I get for merging code on the fly...

manchicken replied:
Happens to all of us :)

manchicken commented:
Thanks for your contribution! I added some comments. Much thanks again!

CTrabant commented:
Thanks.  I think I've addressed most of your comments and I fixed the broken publisher example (suppose I should actually run examples before submitting them to documentation!).

What was probably apparent is that I am not at all fluent in RabbitMQ or AMQP, so I take your word for the suggested changes.
